services/storehost/extStats.go

```diff
@@ -0,0 +1,455 @@
+}
+
+// ExtStatsReporterSetReportInterval updates the report interval (intended for tests)
+func ExtStatsReporterSetReportInterval(interval time.Duration) {
+
+	reportInterval.Store(interval)
+	common.GetDefaultLogger().WithField(`interval`, interval).Info("extStatsReporter: updated report interval")
+}
+
+func NewExtStatsReporter(hostID string, xMgr *ExtentManager, mClient metadata.TChanMetadataService, logger bark.Logger) *ExtStatsReporter {
+
+	return &ExtStatsReporter{
+		hostID:  hostID,
+		xMgr:    xMgr,
+		mClient: mClient,
+		logger:  logger,
+
+		extents: extStatsContextMap{
+			xmap: make(map[string]*extStatsContext),
+		},
+
+		reportPool: sync.Pool{New: func() interface{} { return &report{} }},
+		reportC:    make(chan *report, reportChanBufLen),
+		stopC:      make(chan struct{}),
+	}
+}
+
+func (t *ExtStatsReporter) Start() {
+
+	t.logger.Info("extStatsReporter: started")
+
+	// register callbacks to get notified every time an extent is opened/closed
+	t.xMgr.RegisterCallbacks(t)
+
+	t.wg.Add(1)
+	go t.reporterPump()
+
+	t.wg.Add(1)
+	go t.schedulerPump()
+}
+
+func (t *ExtStatsReporter) Stop() {
+
+	close(t.stopC)
+	t.wg.Wait()
+
+	t.logger.Info("extStatsReporter: stopped")
+}
+
+func (t *ExtStatsReporter) trySendReport(extentID uuid.UUID, ext *extentContext, ctx *extStatsContext) bool {
+
+	// if paused, do nothing
+	if atomic.LoadInt32(&extStatsReporterPaused) == 1 {
+		return false
+	}
+
+	r := t.reportPool.Get().(*report)
+
+	r.ctx = ctx
+	r.extentID = extentID
+	r.timestamp = time.Now().UnixNano()
+
+	// get a snapshot of various extent stats
+	r.firstSeqNum, r.lastSeqNum = ext.getBeginSeqNum(), ext.getLastSeqNum()
+	r.sealed, _ = ext.isSealed()
+
+	// create and send report non-blockingly
+	select {
+	case t.reportC <- r:
+
+		return true // sent
+
+	default:
+		// drop the report, if we are running behind
+		if ctx.lastReport != nil {
+			t.reportPool.Put(ctx.lastReport) // return old 'lastReport' to pool
+		}
+		ctx.lastReport = r // update lastReport to this one
+
+		// t.logger.WithFields(bark.Fields{ // #perfdisable
+		// 	common.TagExt:  extentID,      // #perfdisable
+		// 	`first-seqnum`: r.firstSeqNum, // #perfdisable
+		// 	`last-seqnum`:  r.lastSeqNum,  // #perfdisable
+		// }).Info("extStatsReporter: report dropped") // #perfdisable
+
+		return false // not sent
+	}
+}
+
+func (t *ExtStatsReporter) schedulerPump() {
+
+	defer t.wg.Done()
+
+	interval := reportInterval.Load().(time.Duration)
+	ticker := time.NewTicker(interval)
+
+	t.logger.WithFields(bark.Fields{
+		`report-interval`: interval,
+	}).Info("extStatsReporter: schedulerPump started")
+
+pump:
+	for {
+		select {
+		case <-ticker.C:
+
+			// list of extents that can be deleted
+			deleteExtents := make([]string, 8)
+
+			// get lock shared, while iterating through map
+			t.extents.RLock()
+
+			// pause between extents to spread out the calls
+			pause := time.Duration(interval.Nanoseconds() / int64(1+len(t.extents.xmap)))
```

Comment:
how about rate limiting calls to metadata on the reporterPump and having this pump block on reportC, instead of attempting to spread the load here ? If I understand right, you spread out load is to avoid browning out metadata.. so, say, reportingInteval=1minute and the maxUpdateReplicaStatsPerSecAllowed=10, with the current code, if the number of extents is 600, things will even out nicely. But if the number of extents is 1000, you will exceed the metadata limit.